torch/_inductor/ir.py

```diff
@@ -167,11 +170,21 @@ def get_stride_order(seq: Sequence[int]):
         out[elem] = i
     return out
 
+def _in_faketensormode():
+    from torch.utils._python_dispatch import _get_current_dispatch_mode_stack
+    from torch._subclasses.fake_tensor import FakeTensorMode
+    for m in _get_current_dispatch_mode_stack():
+        if isinstance(m, FakeTensorMode):
+            return True
+    return False
 
 def ir_node_to_tensor(x, guard_shape=True):
     if x is None:
         return None
-    if not guard_shape:
+
+    # for some operators we can not run them in FakeTensorMode (check FallbackKernel.create).
+    # Without FakeTensorMode size/stride have to be int. So we need call size_hint.
+    if not guard_shape or not _in_faketensormode():
```

Comment:
I am still not sure about this. Can you explain this change in more detail?

Reply:
Sure. So we have a fallback for aten._adaptive_avg_pool2d_backward in inductor. When we create a FallbackKernel, we tried to run the op with fake tensors to decide the size/stride of the output. However, the behavior of the op running with fake tensors is not consistent to running in eager mode.

E.g.,  for the inputs to the op:
>  _adaptive_avg_pool2d_backward(Tensor grad_output, Tensor self) -> Tensor

if grad_out is contiguous while self is channels last, eager returns a channels last tensor while in FakeTensorMode we returns a contiguous tensor. This causes vgg16 training to fail when enabling layout optimization.

So I add aten._adaptive_avg_pool2d_backward to the list `fake_incorrect_kernels` in FallbackKernel.create so the op will be run in eager mode to decide the layout of its output.

This cause issue when we enabling dynamic shape though, since the first dimension of the input tensor most likely is batch-size and will be represented as a symbol. torch.empty_strided will fail complaining that it wants size/stride to be all (concrete) int because its not running under a FakeTensorMode.

Reply:
Hol' up

> However, the behavior of the op running with fake tensors is not consistent to running in eager mode.

So fix it (the meta registration). You can consult device hints in the meta function to do different behavior in CPU/CUDA.